consensus/polybft/fsm.go

```diff
@@ -242,13 +237,47 @@ func (f *fsm) Validate(proposal []byte) error {
 		)
 	}
 
-	blockExtra, err := GetIbftExtra(block.Header.ExtraData)
+	currentExtra, err := GetIbftExtra(block.Header.ExtraData)
 	if err != nil {
 		return fmt.Errorf("cannot get extra data:%w", err)
 	}
 
+	parentExtra, err := GetIbftExtra(f.parent.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	currentValidators := f.validators.Accounts()
+	nextValidators := f.validators.Accounts()
+
+	if f.isEndOfEpoch {
+		if err := f.blockBuilder.Reset(); err != nil {
+			return fmt.Errorf("failed to initialize block builder: %w", err)
+		}
+
+		// we need to write commit epoch transaction to intermediate state,
+		// in order to be able to query next validator set
+		if err := f.blockBuilder.WriteTx(block.Transactions[0]); err != nil {
```

Comment:
We are writing `commitEpoch` transaction in the state, so that we are able to query validator set for next epoch. Don't know for alternative solution atm.